extractor/image/image.go

```diff
@@ -0,0 +1,240 @@
+		}
+
+		img.rawSource = rawSource
+		img.src = src
+		return nil
+	}
+	// return only the last error
+	return err
+}
+
+func (img *Image) LayerInfos() ([]imageTypes.BlobInfo, error) {
+	// ignore the cache error
+	layerInfos, _ := img.getLayerInfosInCache(img.name, img.isFile)
+	if layerInfos != nil {
+		return layerInfos, nil
+	}
+
+	// When it doesn't hit a cache, it fetches the image from Docker Engine or Docker Registry
+	if err := img.populateSource(); err != nil {
+		return nil, xerrors.Errorf("failed population: %w", err)
+	}
+
+	layers := img.src.LayerInfos()
+
+	// ignore the cache error
+	_ = img.storeLayerInfosInCache(img.name, img.isFile, layers)
+
+	return layers, nil
+}
+
+func (img Image) getLayerInfosInCache(imageName string, isFile bool) ([]imageTypes.BlobInfo, error) {
+	// tar file doesn't have info in cache
+	if isFile {
+		return nil, nil
+	}
+	rc := img.cache.Get("layerinfos::" + imageName)
+	if rc == nil {
+		return nil, nil
+	}
+	var layerInfos []imageTypes.BlobInfo
+	if err := json.NewDecoder(rc).Decode(&layerInfos); err != nil {
+		return nil, xerrors.Errorf("invalid cache: %w", err)
+	}
+	return layerInfos, nil
+}
+
+func (img Image) storeLayerInfosInCache(imageName string, isFile bool, layers []imageTypes.BlobInfo) error {
+	// it doesn't store the information of a tar file
+	if isFile {
+		return nil
+	}
+
+	b, err := json.Marshal(layers)
+	if err != nil {
+		return xerrors.Errorf("invalid json: %w", err)
+	}
+	return img.cache.SetBytes("layerinfos::"+imageName, b)
+}
+
+func (img *Image) ConfigBlob(ctx context.Context) ([]byte, error) {
+	b, _ := img.getConfigBlobInCache(img.name, img.isFile)
+	if b != nil {
+		return b, nil
+	}
+
+	if err := img.populateSource(); err != nil {
+		return nil, xerrors.Errorf("failed population: %w", err)
+	}
+
+	blob, err := img.src.ConfigBlob(ctx)
+	if err != nil {
+		return nil, xerrors.Errorf("failed to get a config: %w", err)
+	}
+
+	if !img.isFile {
+		_ = img.cache.SetBytes("configblob::"+img.name, blob)
+	}
+
+	return blob, nil
+}
+
+func (img Image) getConfigBlobInCache(imageName string, isFile bool) ([]byte, error) {
+	// tar file doesn't have info in cache
+	if isFile {
+		return nil, nil
+	}
+	rc := img.cache.Get("configblob::" + imageName)
+	if rc == nil {
+		return nil, nil
+	}
+	b, err := ioutil.ReadAll(rc)
+	if err != nil {
+		return nil, xerrors.Errorf("unable to read a cache file: %w", err)
+	}
+	return b, nil
+}
+
+func (img *Image) GetBlob(ctx context.Context, dig digest.Digest) (io.ReadCloser, error) {
+	rc := img.cache.Get(dig.String())
+	if rc != nil {
+		return rc, nil
+	}
+
+	if err := img.populateSource(); err != nil {
+		return nil, xerrors.Errorf("failed population: %w", err)
+	}
+
+	rc, _, err := img.rawSource.GetBlob(ctx, imageTypes.BlobInfo{Digest: dig, Size: -1}, img.blobInfoCache)
+	if err != nil {
+		return nil, xerrors.Errorf("failed to download the layer(%s): %w", dig, err)
+	}
+
+	stream, _, err := compression.AutoDecompress(rc)
```

Comment:
Do we call `Close()` on `stream`? As per the documentation we need too it seems https://github.com/containers/image/blob/master/pkg/compression/compression.go#L130-L134

We should also add tests for this. Currently there aren't any. One of the things to assert in the test would be to ensure `Close()` is eventually called as it seems to be a requirement by the dependency.

Reply:
I added a defer function to clean up io.ReadCloser.
https://github.com/aquasecurity/fanal/pull/70/commits/8672e91ed5f4031ac56bd7fb9fce863fc6804fd7